From 983425e0c34da828a2d64adb137f8b616b62abeb Mon Sep 17 00:00:00 2001
From: David Milburn <dmilburn@redhat.com>
Date: Mon, 5 May 2014 21:04:05 -0400
Subject: [block] nvme: Add getgeo to block ops

Message-id: <1399323848-28227-10-git-send-email-dmilburn@redhat.com>
Patchwork-id: 79398
O-Subject: [RHEL7.1 PATCH BZ 1081734 09/12] NVMe: Add getgeo to block ops
Bugzilla: 1081734
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: Keith Busch <keith.busch@intel.com>

Some programs require HDIO_GETGEO work, which requires we implement
getgeo.

Signed-off-by: Keith Busch <keith.busch@intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
(cherry picked from commit 4cc09e2dc4cbe6009c935b6f12a8376f09124bc5)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/nvme-core.c b/drivers/block/nvme-core.c
index 3def6f3..1236ac6 100644
--- a/drivers/block/nvme-core.c
+++ b/drivers/block/nvme-core.c
@@ -25,6 +25,7 @@
 #include <linux/errno.h>
 #include <linux/fs.h>
 #include <linux/genhd.h>
+#include <linux/hdreg.h>
 #include <linux/idr.h>
 #include <linux/init.h>
 #include <linux/interrupt.h>
@@ -1796,12 +1797,22 @@ static void nvme_release(struct gendisk *disk, fmode_t mode)
  kref_put(&dev->kref, nvme_free_dev);
 }
 
+static int nvme_getgeo(struct block_device *bd, struct hd_geometry *geo)
+{
+ /* some standard values */
+ geo->heads = 1 << 6;
+ geo->sectors = 1 << 5;
+ geo->cylinders = get_capacity(bd->bd_disk) >> 11;
+ return 0;
+}
+
 static const struct block_device_operations nvme_fops = {
  .owner  = THIS_MODULE,
  .ioctl  = nvme_ioctl,
  .compat_ioctl = nvme_compat_ioctl,
  .open  = nvme_open,
  .release = nvme_release,
+ .getgeo  = nvme_getgeo,
 };
 
 static void nvme_resubmit_bios(struct nvme_queue *nvmeq)
-- 
1.7.1